src/coprocessor/dag/batch_executor/table_scan_executor.rs

```diff
@@ -0,0 +1,906 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use cop_datatype::{EvalType, FieldTypeAccessor};
+use kvproto::coprocessor::KeyRange;
+use tipb::expression::FieldType;
+use tipb::schema::ColumnInfo;
+
+use crate::storage::Store;
+use crate::util::collections::HashMap;
+
+use super::interface::*;
+use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+use crate::coprocessor::dag::expr::{EvalConfig, EvalContext};
+use crate::coprocessor::dag::Scanner;
+use crate::coprocessor::Result;
+
+pub struct BatchTableScanExecutor<S: Store>(
+    super::scan_executor::ScanExecutor<
+        S,
+        TableScanExecutorImpl,
+        super::ranges_iter::PointRangeEnable,
+    >,
+);
+
+impl<S: Store> BatchTableScanExecutor<S> {
+    pub fn new(
+        store: S,
+        config: Arc<EvalConfig>,
+        columns_info: Vec<ColumnInfo>,
+        key_ranges: Vec<KeyRange>,
+        desc: bool,
+    ) -> Result<Self> {
+        let is_column_filled = vec![false; columns_info.len()];
+        let mut key_only = true;
+        let mut handle_index = None;
+        let mut schema = Vec::with_capacity(columns_info.len());
+        let mut columns_default_value = Vec::with_capacity(columns_info.len());
+        let mut column_id_index = HashMap::default();
+
+        for (index, mut ci) in columns_info.into_iter().enumerate() {
+            schema.push(super::scan_executor::field_type_from_column_info(&ci));
+            columns_default_value.push(ci.take_default_val());
+
+            if ci.get_pk_handle() {
+                handle_index = Some(index);
+            } else {
+                key_only = false;
```

Comment:
why?

Reply:
The comment of the key_only is:

```rust
    /// Whether or not KV value can be omitted.
    ///
    /// It will be set to `true` if only PK handle column exists in `schema`.
```

So when we meet a non-PK handle column, `key_only` is `false`.

Reply:
Could we set it after this loop? Maybe just check whether the column_id_index is empty? So we needn't set it in the loop which will confuse the reviewers

Reply:
But how can we determine it outside the loop?

Reply:
BTW I have supplied some comments about what 4 kind of things we need to do in this loop. Do you accept this solution?

Reply:
ok